### ISSUE TITLE
Load/Save Layout also saves/overwrites persistent data like "recent items"

### ISSUE THREAD

**Issue by MBilderbeek** [openMSX/openMSX#1698]
See title. The "layout file" is actually a full `imgui.ini` persistence file, where much more is stored than just the layout. So when loading one, you overwrite your 'recent items' (and possibly other saved data that is not necessary layout related).
Also when trying to start with a fresh layout, you can now only remove the full `imgui.ini` file.

I suggest we should be able to save/load/reset a specific set of ini items instead of always the full `imgui.ini`.

**Comment by m9710797** [comment]
Or alternatively, save the extra stuff in a different file than "imgui.ini".

**Comment by MBilderbeek** [comment]
Looks like we could fix several issues if we would fix this issue.

So, let's start designing.

I guess we can keep the imgui.ini to save layouts.
And we should move all our own custom persistency to another file. What file name and what format do we use? (The ini was only because of `imgui.ini`, so we're definitely not bound to that... in openMSX we usually use XML.)

**Comment by m9710797** [comment]
To be clear we should not move _all_ our custom additions in `imgui.ini` to another file. Maybe even the majority can remain? Some examples:
* Most windows remember if they are open/closed. Maybe this info should remain in `imgui.ini`: it should be restored via `main menu bar > Settings > GUI > Restore Layout`. All that option does is load a different Dear ImGui .ini files, so all info that should change on `Restore Layout` should remain in `imgui.ini`.
* Various stuff to configure the layout, like CPUFlags horizontial/vertical, the number of columns in the hex editor(s), the various OSD icons and reversebar settings, the status bar, ... These should all remain in `imgui.ini`.
* Recently used items, like the last N used machines, roms/disk images, ... You do want to keep these when switching to a new layout. So these do not belong in `imgui.ini`. Note that the console history is somewhat similar, that one is stored ad-hoc in `~/.openMSX/persistent/console/history.txt`. I'm not saying this was the best idea and that we should now do the same, but at least there is some precedent.
* Maybe some stuff can be moved to `settings.xml`? E.g. the behavior of line-wrapping in the console can be configured (via right-click context menu). I assume you want this setting fixed, even if you change layout? Though for other (and maybe even this) settings this choice is not always clear.
* I think we once had a request to be able to save/restore the debugger session independent from the GUI layout (and independent from `settings.xml`). So stuff like which breakpoints/watchpoints are set. The easiest way to make this independent is to store it in a separate file. But it's not always clear what exactly belongs to the debugger session and what belongs to the layout. E.g. which hex-editors are open, is that layout or debugger session?

---

I believe before we can solve this we should first get a (reasonable) idea about how we want to tackle "session management". There are various aspects:
* Layout stuff (`imgui.ini`).
* Generic settings (now stored in `settings.xml`).
* The recently added "machine setup" stuff.
* Persistent storage (like SRAM content).
* Recently used stuff (like console history, last used machines, ...). I think it's fine if the user has no control over this?
* Do we want an independent new category for a debugger session?

Personally I don't know how this all should fit together. Should it all be independent stuff, or do we want to couple (some of) these? Maybe we should not fully answer this now, but it can have an influence on the decision of where (in what file) to store what information.

**Comment by MBilderbeek** [comment]
Let's take this piece by piece. For instance,start with something that is already clear: the recent items. 
A proposal could be to store them for each type in a separate file (for simplicity), in the folder `~/.openMSX/persistent/gui/recent`.

Current recent items in `imgui.ini` (with some examples):
- `[openmsx][machine]
machine.recent=Nikko_PC-70100` (originating from `ImGuiMachine.cc`)
- `[openmsx][media]` with all media and extensions like (originating from `ImGuiMedia.cc`):
  - `diska.image.recent.name` (and all `diskX`)
  - `carta.rom.recent.name` (and all `cartX`)
  - `carta.extension.recent.name` (and all `cartX`)
  - `hda.recent.name` (and all `hdX`)
  - `cassette.recent.name`
  - `laserdisc.recent.name`
  - `extension.recent.name`

So, we could put them in separate files like `machine.txt`, `diska.txt`, `cassette.txt`, `carta.txt`, `cartb.txt`, etc.

**Comment by m9710797** [comment]
Good start, but I think this is too simple. For example for roms/disks we also store the list of applied IPS patches. For roms also the rom-type.

**Comment by MBilderbeek** [comment]
OK, so let's see how we can incorporate this. E.g. we could make a tab-separated list of fields in the file:

For disk images:
`filename<tab>patch1<tab>patch2<tab>...<tab>patchN`
for ROM images:
`filename<tab>mappertype<tab>patch1<tab>patch2<tab>...<tab>patchN`
and for all others:
`filename`

Using `<tab>` because no filename, patch file or mapper type will have a tab character in it, so no escaping needed.

I realize this only works because there are no other variable length data items to be put in the text file... but it is very simple. Good enough?

**Comment by MBilderbeek** [comment]
A Tcl dict could also work, as you wrote on IRC. And is a little more flexible. And we have Tcl anyway.
You wrote: `carta myrom.rom patches {} romtype Konami`
I'm fine with that. You also wrote "we could combine all recent items into a single file". How? By making it a bit Tcl list per "item" (=carta, cartb, diska, diskb, etc.)?

**Comment by m9710797** [comment]
Yes, we need some kind of string-escaping. Tcl has a solution for that, and we already use it in various places in openMSX, so we can use it here as well.
And instead of a plain list I propose to add some more (but still minimal) structure. So a list of key-value pairs (a Tcl dict).
I didn't explicitly say this (on IRC), but the value corresponding to the `patches` key could be a list instead of a single file. And this all works even with filenames that have embedded space/tab characters.

The first key in the list, `carta` in my example, already indicates the group. So not strictly needed to make lists per item. But that's of course another possibility. We could also use XML for even more structure. It doesn't really matter much.

**Comment by MBilderbeek** [comment]
Well, we have to choose something in order to proceed...
